_drafts/PR-blogs/planning-docs/example-code/tic-tac-toe/original/script.js

```diff
@@ -0,0 +1,87 @@
+const statusDisplay = document.querySelector('.game--status');
+
+let gameActive = true;
+let currentPlayer = "X";
+let gameState = ["", "", "", "", "", "", "", "", ""];
+
+const winningMessage = () => `Player ${currentPlayer} has won!`;
+const drawMessage = () => `Game ended in a draw!`;
+const currentPlayerTurn = () => `It's ${currentPlayer}'s turn`;
+
+statusDisplay.innerHTML = currentPlayerTurn();
+
+const winningConditions = [
+    [0, 1, 2],
+    [3, 4, 5],
+    [6, 7, 8],
+    [0, 3, 6],
+    [1, 4, 7],
+    [2, 5, 8],
+    [0, 4, 8],
+    [2, 4, 6]
+];
+
+function handleCellPlayed(clickedCell, clickedCellIndex) {
+    gameState[clickedCellIndex] = currentPlayer;
+    clickedCell.innerHTML = currentPlayer;
+}
+
+function handlePlayerChange() {
+    currentPlayer = currentPlayer === "X" ? "O" : "X";
+    statusDisplay.innerHTML = currentPlayerTurn();
+}
+
+function handleResultValidation() {
+    let roundWon = false;
+    for (let i = 0; i <= 7; i++) {
+        const winCondition = winningConditions[i];
+        let a = gameState[winCondition[0]];
+        let b = gameState[winCondition[1]];
+        let c = gameState[winCondition[2]];
+        if (a === '' || b === '' || c === '') {
+            continue;
+        }
+        if (a === b && b === c) {
+            roundWon = true;
+            break
+        }
+    }
+
+    if (roundWon) {
+        statusDisplay.innerHTML = winningMessage();
+        gameActive = false;
+        return;
+    }
+
+    let roundDraw = !gameState.includes("");
+    if (roundDraw) {
+        statusDisplay.innerHTML = drawMessage();
+        gameActive = false;
+        return;
+    }
+
+    handlePlayerChange();
+}
```

Comment:
Firstly, some things I really like about this code: 
- It's wrapped up in a discrete function with a good name
- It leverages early returns, and doesn't tangle its logic in a bunch of `if`/`else` trees, which can be hard to refactor down the line
- It has nice whitespace to separate out the different parts, which helps us focus on the individual pieces of the function.

There are some things we can do which will make it even more readable: 
- Refactoring your `for` loop to a `reduce` function. `for` loops are good for looping tasks which require mutating state. In this context, we're checking the current board against the game's winning conditions to see whether to continue the game or whether a win or draw has occurred. This is a "data-driven" activity, and `reduce` is a tool which is much better suited for data-driven tasks. 
- Refactoring some boolean logic to use strings which represent game states. 
- Giving names to the logic checks you're performing, which will help us to more easily understand what exactly constitutes a win vs. a draw. When you've got variables to represent concepts, it makes things easier to read, seeing as you can rely on the variable name to communicate what's going on. Without a variable name, you have have to read the logic itself to understand what's going on. The logic in this context is not that hard to understand, but it nevertheless takes more time and effort to grok than a named variable would. 

So, that being said, here's how I would refactor the code: 

```js
// At the top of the file, we can define some handy constants which represent game states: 
const IN_PROGRESS = "in progress"
const WIN = "win"
const DRAW = "draw"

// the refactored function itself
function handleResultValidation() {
    const gameStatus = winningConditions.reduce((status, winCondition) => {
        const a = gameState[winCondition[0]]
        const b = gameState[winCondition[1]]
        const c = gameState[winCondition[2]]
        
        // a "vector" means either a column or a row
        const incompleteVector = (a === '' && b === '' && c === '')
        const filledVector = (a === b && b === c)
        const boardFilledUp = !gameState.includes("") 
        const gameFinished = filledVector && !incompleteVector

        if (status === WIN) {
            return WIN
        }

        if (boardFilledUp || status === DRAW) {
            return DRAW
        }
        
        return gameFinished ? WIN : IN_PROGRESS;
    }, IN_PROGRESS)

    return gameStatus === IN_PROGRESS ? handlePlayerChange() : endGame(gameStatus)
}
```